File: backend-django/events/management/commands/send_event_notifications.py
from datetime import timedelta

from django.core.management.base import BaseCommand
from django.utils import timezone

from emails import (
    send_event_access_revealed,
    send_event_organizer_digest,
    send_event_reminder,
)
from events.models import Event, EventStatus
from registrations.models import RegistrationStatus


class Command(BaseCommand):
    help = (
        "Envoie les notifications planifiées liées aux événements "
        "(rappels participants, révélations d'accès et récapitulatifs organisateur)."
    )

    def handle(self, *args, **options):
        now = timezone.now()
        upcoming_events = (
            Event.objects
            .filter(status=EventStatus.PUBLISHED, date_end__gt=now)
            .select_related('company')
            .prefetch_related('registrations__participant')
        )

        reminder_count = 0
        reveal_count = 0
        digest_count = 0

        for event in upcoming_events:
            confirmed_registrations = [
                reg for reg in event.registrations.all()
                if reg.status == RegistrationStatus.CONFIRMED
            ]
            delta = event.date_start - now
            update_fields = []

            reminder_key = None
            reminder_field = None
            if timedelta(0) < delta <= timedelta(hours=3) and event.reminder_3h_sent_at is None:
                reminder_key = '3h'
                reminder_field = 'reminder_3h_sent_at'
            elif timedelta(hours=3) < delta <= timedelta(days=1) and event.reminder_1d_sent_at is None:
                reminder_key = '1d'
                reminder_field = 'reminder_1d_sent_at'
            elif timedelta(days=1) < delta <= timedelta(days=7) and event.reminder_7d_sent_at is None:
                reminder_key = '7d'
                reminder_field = 'reminder_7d_sent_at'

            if reminder_key and confirmed_registrations:
                for registration in confirmed_registrations:
                    send_event_reminder(registration, reminder_key)
                    reminder_count += 1
                setattr(event, reminder_field, now)
                update_fields.append(reminder_field)

            reveal_targets = []
            if (
                event.format in ['ONSITE', 'HYBRID']
                and event.address_visibility == 'PARTIAL'
                and event.address_reveal_date
                and now >= event.address_reveal_date
                and event.address_reveal_email_sent_at is None
                and event.address_full
            ):
                reveal_targets.append('address')
            if (
                event.format in ['ONLINE', 'HYBRID']
                and event.online_visibility == 'PARTIAL'
                and event.online_reveal_date
                and now >= event.online_reveal_date
                and event.online_reveal_email_sent_at is None
                and event.online_link
            ):
                reveal_targets.append('online')

            if reveal_targets and confirmed_registrations:
                for registration in confirmed_registrations:
                    send_event_access_revealed(registration, reveal_targets)
                    reveal_count += 1
                if 'address' in reveal_targets:
                    event.address_reveal_email_sent_at = now
                    update_fields.append('address_reveal_email_sent_at')
                if 'online' in reveal_targets:
                    event.online_reveal_email_sent_at = now
                    update_fields.append('online_reveal_email_sent_at')

            if (
                timedelta(0) < delta <= timedelta(days=1)
                and event.organizer_digest_sent_at is None
            ):
                send_event_organizer_digest(event)
                digest_count += 1
                event.organizer_digest_sent_at = now
                update_fields.append('organizer_digest_sent_at')

            if update_fields:
                event.save(update_fields=sorted(set(update_fields)))

        self.stdout.write(
            self.style.SUCCESS(
                "Notifications envoyées "
                f"(rappels={reminder_count}, revelations={reveal_count}, recaps={digest_count})"
            )
        )

File: backend-node/src/scripts/sendEventNotifications.js
'use strict';

/**
 * sendEventNotifications.js — Notifications planifiées liées aux événements.
 *
 * Équivalent Node.js du management command Django :
 *   python manage.py send_event_notifications
 *
 * Ce script envoie :
 *   - Rappels aux participants confirmés (D-7, D-1, H-3)
 *   - Révélations d'accès (adresse complète et/ou lien visio)
 *   - Récapitulatif organisateur (J-1)
 *
 * Usage :
 *   node src/scripts/sendEventNotifications.js
 *
 * À exécuter via cron (ex: toutes les heures) :
 *   0 * * * * node /app/src/scripts/sendEventNotifications.js
 */

require('dotenv').config();

const { Op } = require('sequelize');
const { Event, Registration } = require('../models');
const {
  sendEventReminder,
  sendEventAccessRevealed,
  sendEventOrganizerDigest,
} = require('../services/emailService');

const MS_PER_HOUR = 60 * 60 * 1000;
const MS_PER_DAY = 24 * MS_PER_HOUR;

async function run() {
  const now = new Date();

  const upcomingEvents = await Event.findAll({
    where: {
      status: 'PUBLISHED',
      date_end: { [Op.gt]: now },
    },
    include: [
      {
        association: 'company',
        attributes: ['id', 'company_name', 'recovery_email'],
      },
      {
        association: 'registrations',
        include: [
          { association: 'participant', attributes: ['id', 'first_name', 'last_name', 'email'] },
        ],
      },
    ],
  });

  let reminderCount = 0;
  let revealCount = 0;
  let digestCount = 0;

  for (const event of upcomingEvents) {
    const delta = new Date(event.date_start) - now; // milliseconds

    const confirmedRegistrations = (event.registrations || []).filter(
      r => r.status === 'CONFIRMED'
    );

    // ─── Rappels participants ─────────────────────────────────────────────────
    let reminderKey = null;
    let reminderField = null;

    if (delta > 0 && delta <= 3 * MS_PER_HOUR && event.reminder_3h_sent_at === null) {
      reminderKey = '3h';
      reminderField = 'reminder_3h_sent_at';
    } else if (delta > 3 * MS_PER_HOUR && delta <= MS_PER_DAY && event.reminder_1d_sent_at === null) {
      reminderKey = '1d';
      reminderField = 'reminder_1d_sent_at';
    } else if (delta > MS_PER_DAY && delta <= 7 * MS_PER_DAY && event.reminder_7d_sent_at === null) {
      reminderKey = '7d';
      reminderField = 'reminder_7d_sent_at';
    }

    if (reminderKey && confirmedRegistrations.length > 0) {
      for (const reg of confirmedRegistrations) {
        // Attacher l'event à la registration pour l'email
        reg.event = event;
        await sendEventReminder(reg, reminderKey).catch(() => {});
        reminderCount++;
      }
      await event.update({ [reminderField]: now });
    }

    // ─── Révélations d'accès ──────────────────────────────────────────────────
    const revealTargets = [];

    if (
      ['ONSITE', 'HYBRID'].includes(event.format) &&
      event.address_visibility === 'PARTIAL' &&
      event.address_reveal_date &&
      now >= new Date(event.address_reveal_date) &&
      event.address_reveal_email_sent_at === null &&
      event.address_full
    ) {
      revealTargets.push('address');
    }

    if (
      ['ONLINE', 'HYBRID'].includes(event.format) &&
      event.online_visibility === 'PARTIAL' &&
      event.online_reveal_date &&
      now >= new Date(event.online_reveal_date) &&
      event.online_reveal_email_sent_at === null &&
      event.online_link
    ) {
      revealTargets.push('online');
    }

    if (revealTargets.length > 0 && confirmedRegistrations.length > 0) {
      for (const reg of confirmedRegistrations) {
        reg.event = event;
        await sendEventAccessRevealed(reg, revealTargets).catch(() => {});
        revealCount++;
      }
      const revealUpdates = {};
      if (revealTargets.includes('address')) revealUpdates.address_reveal_email_sent_at = now;
      if (revealTargets.includes('online')) revealUpdates.online_reveal_email_sent_at = now;
      await event.update(revealUpdates);
    }

    // ─── Récapitulatif organisateur (J-1) ─────────────────────────────────────
    if (delta > 0 && delta <= MS_PER_DAY && event.organizer_digest_sent_at === null) {
      await sendEventOrganizerDigest(event).catch(() => {});
      digestCount++;
      await event.update({ organizer_digest_sent_at: now });
    }
  }

  console.log(
    `Notifications envoyées (rappels=${reminderCount}, revelations=${revealCount}, recaps=${digestCount})`
  );
}

run()
  .then(() => process.exit(0))
  .catch(err => {
    console.error('Erreur lors de l\'envoi des notifications:', err);
    process.exit(1);
  });
